wile/argtypes.py

```diff
@@ -5,25 +5,39 @@
 
 import click
 
-_DomainWebrootTuple = namedtuple('DomainWebrootTuple', ['domain', 'webroot'])
+_DomainRemoteWebrootTuple = namedtuple('DomainRemoteWebrootTuple', ['domain', 'remote', 'webroot'])
 
 
-class _DomainWebrootType(click.ParamType):
+class _DomainRemoteWebrootType(click.ParamType):
     domain = None
+    remote = None
     webroot = None
 
     def convert(self, value, param, ctx):
-        if isinstance(value, _DomainWebrootTuple):
+        if isinstance(value, _DomainRemoteWebrootTuple):
             return value
         url = value.split(':')
-        if len(url) not in (1, 2):
-            self.fail('could not parse %s as DOMAIN[:WEBROOT]' % value)
+        if len(url) not in range(1, 5):
+            self.fail('could not parse %s as DOMAIN:[[[USER@]HOST[:PORT]:]WEBROOT]' % value)
         domain = url[0]
-        webroot = len(url) > 1 and os.path.expanduser(url[1]) or None
-        return _DomainWebrootTuple(domain=domain, webroot=webroot)
+        if len(url) > 2:
+            split_tmp = url[1].split('@')
+            if url[2].isdigit():
+                if len(split_tmp) > 1:
+                    remote = (split_tmp[0], split_tmp[1], url[2])
+                else:
+                    remote = (None, split_tmp[0], url[2])
+                webroot = len(url) == 4 and os.path.expanduser(url[3]) or None
+            else:
+                remote = (split_tmp[0], split_tmp[1], None)
+                webroot = os.path.expanduser(url[2])
+        else:
```

Comment:
This block seems a bit error-prone and not very readable.
I'd suggest leaving `_DomainWebrootType` untouched (besides changing the first split to `value.split(':', 1)`) and passing the `webroot` - including its remote prefix - directly on to `_generate_domain_and_webroot_lists_from_args`. There you can break it apart with a regex and create a new type (maybe `argtypes.RemotePathType` ?). Lastly, you do type-checking when storing, instead of passing the extra `remote` argument. 

This regex seems to work and also falls-back nicely:
```python
r'^(?:(?:(?P<remote_user>[^@]+)@)?(?P<remote_host>[^@:]+)(?::(?P<remote_port>[0-9]+))?:)?(?P<webroot>[^:]+)$'
```
(though it's hard to argue for its simplicity... :grin: but I think it only looks bad because of the named matches)

This way we minimize the places we have to touch and keep the complexity to the one regex, avoiding the passing around of unnamed tuples.